poo/design_patterns/en/OBSERVER_PATTERN.md

```diff
@@ -0,0 +1,232 @@
+
+## So, what do we do Einstein ?
+
+The **Observer Pattern** is born for this kind of situation.
+
+This pattern is made of 2 subjects, like said earlier : an **observable** subject and one or more **observer** subject.
+
+We'll use the abstraction principle to write a reusable code in most or all cases.
+
+```java
+interface Observer {
+
+}
+
+interface Observable {
+
+}
+```
+
+These interfaces seem very empty
+
+With this pattern, a class so-called **Observer** must be notifiable at any time of the day and night and an **observable** class must notify each one of the elements which are observing it.
+
+Let's add these 2 methods :
+
+```java
+interface Observer {
+    void notify();
+}
+
+interface Observable {
+    void notifyObservers();
+}
+```
+
+Okey </br> Of course we must be able to link an **Observer** to an **Observable** element :
+
+```java
+interface Observer {
+    void notify();
+}
+
+interface Observable {
+    void notifyObservers();
+    void addObserver(Observer observer);
+}
+```
+
+Let's try to rewrite our alarm system making use of these 2 interfaces :
+
+```java
+class Alarm implements Observable {
+    
+    private final List<Observer> observers = new ArrayList<>();
+    
+    public void alert() {
+        notifyObservers();
+        this.speaker.startAlarm();
+    }
+
+    @Override
+    public void addObserver(Observer observer) {
+        observers.add(observer);
+    }
+    
+    @Override
+    public void notifyObservers() {
+        observers.foreach(Observer::notify);
+    }
+}
+
+class PoliceCallerComponent implements Observer {
+    
+    @Override 
+    public void notify() {
+        System.out.println("Police successfully called");
+    }
+}
+
+class DoorLockerComponent implements Observer {
+
+    @Override
+    public void notify() {
+        lockAllDoors();
+        System.out.print("The thief is blocked inside");
+    }
+
+    private void lockAllDoors() {
+        //...
+    }
+
+}
+
+//...
+```
+
+*I'm not going to write all of the systems, you understood the main idea.*
+
+You can't figure yourself what the method `notifyObservers()` will produce ?
+Go read the course about the [functional interfaces](https://github.com/readthedocs-fr/notions/blob/master/java/interfaces_fonctionnelles/en/FUNCTIONAL_INTERFACES.md) as well as the note about the [Method Referencing](https://github.com/readthedocs-fr/notions/blob/master/java/interfaces_fonctionnelles/en/FUNCTIONAL_INTERFACES.md#toujours-plus-court-toujours-plus-loin).
+
+This line will produce the same effect as the following for loop :
+       
+```java
+    for(Observer observer : observers) {
+        observer.notify();
+    }
+```
+
+We now have a code that perfectly suits the **Observer Pattern**. Let's add a control center for the realistic touch and it will surely please to our boss.
```

Comment:
```suggestion
We now have a code that perfectly suits the **Observer Pattern**. Let's add a control center for the realistic touch and it will surely please our boss.
```